app/models/user.rb

```diff
@@ -87,20 +85,38 @@ def self.next_membership_payment_date(user_id)
     next_membership_payment_dates(user_id).first
   end
 
-
   # TODO this should not be the responsibility of the User class. Need a MembershipManager class for this.
   def self.next_membership_payment_dates(user_id)
     next_payment_dates(user_id, THIS_PAYMENT_TYPE)
   end
 
-
-  # ----------------------------------
+  def self.clear_all_proof_of_membership_jpg_caches
+    all.each do |user|
+      user.clear_proof_of_membership_jpg_cache
+    end
+  end
 
   after_update :clear_proof_of_membership_jpg_cache,
                if: Proc.new { saved_change_to_member_photo_file_name? ||
-                              saved_change_to_first_name? ||
-                              saved_change_to_last_name? ||
-                              saved_change_to_membership_number? }
+                 saved_change_to_first_name? ||
+                 saved_change_to_last_name? ||
+                 saved_change_to_membership_number? }
+
+  # @return [ActiveSupport::Duration]
+  def self.membership_expired_grace_period
+    ActiveSupport::Duration.days(AdminOnly::AppConfiguration.config_to_use.membership_expired_grace_period.to_i)
+  end
+
+  def self.most_recent_upload_method
+    MOST_RECENT_UPLOAD_METHOD
+  end
+
+  # @return [ActiveSupport::Duration]
+  def self.days_can_renew_early
+    ActiveSupport::Duration.days(AdminOnly::AppConfiguration.config_to_use.payment_too_soon_days.to_i)
```

Comment:
Same comment as above.

Reply:
Also, this method (and the similar above, and maybe others here in `User` class) don't seem to be belong here, since they are not attributes of, nor rely on, User class or user instance properties.

Perhaps eventually in a `Membership` class(?) - but for now I'd suggest `AppConfiguration`.

Reply:
Agree they ultimately need to be refactored out to a different class.
(I've started work on refactoring.)

They _are_ getting their values from AppConfiguration.

Reply:
@weedySeaDragon I had meant that the method `days_can_renew_early` might be better placed in `AppConfiguration`.  This makes sense, as that class could be imbued with "knowledge" of what the settings represent.

Reply:
The instance method `days_can_renew_early` gets its result from the class method of the same name.  The class method ultimately gets the info from AppConfiguration, which is responsible for providing the definitive values about application settings/configurations.

The class method is the public (external) portal for requesting the info from other sources (classes/instances, whatever).  Since this is a class-level piece of information (e.g. it doesn't vary with instances), I have a User class method responsible for it.

I totally agree that this does not belong in the User class.  But for now, in order to get renewals implemented, I'm putting it here along with all of the other membership term related methods that have piled up in the class.  The User class needs this information in order to do membership term (date) calculations.

We should definitely review what the appropriate responsibility is for the AppConfiguration class, as well as the User class and more.  But I don't want those larger, more involved (and overdue) conversations to derail this right now.

It'd be helpful if you opened a task in Pivotal Tracker to "define the responsibility of the AppConfiguration class" and recorded your thoughts there.  That'd be a good starting place.